testing_and_setup/compass/manage_regression_suite.py

```diff
@@ -193,19 +196,276 @@ def process_test_clean(test_tag, work_dir):  # {{{
+    local_parallel_code += "           number_of_procs += background_process[1] \n"
+    local_parallel_code += "           if index < len(procs)-1: \n"
+    local_parallel_code += "#             print('STILL HAVE MORE TO PROCESS') \n"
+    local_parallel_code += "             if number_of_procs >= procs[index]: \n"
+    local_parallel_code += "               continue_add = True \n"
+    local_parallel_code += "           else: \n"
+    local_parallel_code += "             if len(Queue_running) == 0: \n"
+    local_parallel_code += "               Done = True \n"
+    local_parallel_code += "               continue_add = False \n"
+    local_parallel_code += "     elif Done and not continue_add: \n"
+    local_parallel_code += "      return \n"
+    local_parallel_code += " \n"
+    local_parallel_code += "  return\n"
+    local_parallel_code += "print('--- PROCESSING PREREQS ---')\n"
+    local_parallel_code += "stream_queue(prereq_commands, number_of_procs)\n"
+    local_parallel_code += "print('--- PROCESSING commands ---')\n"
+    local_parallel_code += "stream_queue(allocation_commands, number_of_procs)\n"
+    local_parallel_code += "end_time = time.time()\n"
+    local_parallel_code += "print('parallel run time: {} min'.format((end_time - start_time) / 60))\n"
+    local_parallel_code = write_regression_script_testcase_data_bottom(
+        local_parallel_code)
+
+    return local_parallel_code
+    # }}}
+
+
+def write_regression_local_parallel_top(work_dir, suite_tag, nodes):
+    # {{{
+    local_parallel_code = "#!/usr/bin/env python\n\n\n"
+    local_parallel_code += "import time\n"
+    local_parallel_code += "import psutil\n"
+    local_parallel_code += "import sys\n"
+    local_parallel_code += "import os\n"
+    local_parallel_code += "import subprocess\n"
+    local_parallel_code += "import numpy as np\n"
+    local_parallel_code += "import multiprocessing as mp\n"
+    local_parallel_code += "out, err = subprocess.Popen(['nproc'],stdout=subprocess.PIPE, stderr=subprocess.STDOUT).communicate()\n"
+    local_parallel_code += "nodes = {}\n".format(nodes)
+    local_parallel_code += "number_of_procs = nodes * int(out.split()[0])\n"
+    local_parallel_code += "print('Number of usable Processors: {}'.format(number_of_procs))\n"
+    local_parallel_code += "os.environ['PYTHONUNBUFFERED'] = '1'\n"
+    local_parallel_code += "test_failed=False\n"
+    local_parallel_code += "if not os.path.exists('case_outputs'):\n"
+    local_parallel_code += "    os.makedirs('case_outputs')\n"
+    local_parallel_code += "base_path = '" + work_dir + "'\n"
+    local_parallel_code += "os.chdir(base_path)\n"
+    local_parallel_code += "locations = []\n"
+    local_parallel_code += "procs = []\n"
+    local_parallel_code += "run_file_names = []\n"
+    local_parallel_code += "prereq_commands = []\n"
+    local_parallel_code += "allocation_commands = []\n"
+    local_parallel_code += "commands = []\n\n\n"
```

Comment:
Whenever you find yourself making several (in this case 6) lists in python that are all the same length, it is a good sign that a dictionary of lists, a list of dictionaries, or a dictionary of dictionaries would likely be a better approach.  In this case, I would suggest a dictionary of dictionaries.  The outer keys would be the names of the tests.  The inner names would be the the current names of the lists (possibly made shorter).  This would save you a lot of appending and indexing that is not easy to read.  The indices are also arbitrary and not meaningful.

Reply:
this will take a hot min to restructure

Reply:
@xylar  last commit should have fixed it up a bit